File: java/goldbach.java
//Program for computing Goldbach's Weak Conjecture for an entered value
//For any odd integer greater than seven, finds triplet of odd primes that sums to that integer
//By Adrian Faircloth
//9-9-2021
import java.util.Scanner;
import java.util.Arrays;
public class goldbach {
	
private static int[] primes;	
private static int[] triplet = new int[3];
	public static void main(String[] args) 
	{
	//Takes user input, checks if input meets Goldbach requirements
	//If yes, finds triplet of primes w/ smallest vector norm whose sum is input value,
	//	sorts triplet and prints
	//If no, prints error message and exits
		int inputVal;
		boolean meetsReqs;
		Scanner in = new Scanner(System.in);
		System.out.print("Enter an odd integer greater than seven: ");
		inputVal = in.nextInt();
		meetsReqs = checkReqs(inputVal);
		if (meetsReqs == true)
		{
			primes = findOddPrimes(inputVal);
			findTriplet(inputVal);
			System.out.println(inputVal + ": (" + triplet[0] + ", " + triplet[1] + ", " + triplet[2] + ")"); 
		}
		else
		{
			System.out.println("Invalid input (not odd integer greater than seven)");	
		}
	}	

	public static boolean checkReqs(int val) 
	{ 
	//Checks if a value satisfies the requirements for Goldbach's weak conjecture (odd int > 7)
		if(val % 2 == 1 && val > 7)
		{ 
			return true;
		}
		else
			return false;
	}
	
	public static int[] findOddPrimes(int n) 
	{
	//Creates boolean array isPrime[] of all values < n
	//Any value in isPrime that is not an odd  prime is set to false
		boolean isPrime[] = new boolean[n];
		for (int i = 1; i < n; i++)
			isPrime[i] = true;

		for (int i = 2; i*i <= n; i++) 
		{
			if (isPrime[i] == true) 
			{
				for (int j = i*i; j < n; j += i)
					isPrime[j] = false;
			}
		}
		isPrime[2] = false;	//Removing 2 from set of primes	
		
	//Counting number of odd primes in range < n
	//and storing in an int array
		int numPrimes = 0;
		for (int i = 0; i < n; i++) 
		{
			if (isPrime[i] == true)
				numPrimes++;
		}
		int[] primeVals = new int[numPrimes];
		int pointer = 0;
		for (int i = 0; i < n; i++) 
		{
			if (isPrime[i] == true) 
			{
				primeVals[pointer] = i;
				pointer++;
			}
		}
		return primeVals;
	}
	
	public static boolean isPrime(int val) 
	{
	//Checks if a value is in the set of primes < n
		for (int i = 2; i <= Math.sqrt(val); i++)
		{
			if (val % i == 0)
				return false;
		}
		return true;
	}
	
	public static void findTriplet(int sum) 
	{
	//Finds the triplet of unique odd primes that sum to an entered value
	//and has the smallest vector norm
	//Compares the norms of each possible triplet of primes, stores triplet w/ smaller norm each comparison
	//Finally, triplet will contain values with smallest vector norm
	//In the case of a tie, the triplet with the larger smallest value is returned
		double smallestNorm = 0;	
		double currNorm = 0;
		int k = 0;
		int i = 0;
		int j = 0;
		int[] tempTrip = new int[3];
		for (i = 1; i < sum; i++)
		{
			for (j = 1; j < sum ; j++)
			{
				if(isPrime(i) == true && isPrime(j) == true)
				{
					k = sum - (j + i);
					if  (isPrime(k) && k != j && k != i && j != i)
					{
						currNorm = Math.sqrt((i*i) + (j*j) + (k*k));
						if (smallestNorm == 0 || currNorm < smallestNorm)
						{
							triplet[0] = i;
							triplet[1] = j;
							triplet[2] = k;
							smallestNorm = currNorm;
							Arrays.sort(triplet);
        
						}	
						else if (smallestNorm == currNorm)
						{
							tempTrip[0] = i;
							tempTrip[1] = j;
							tempTrip[2] = k;
							Arrays.sort(tempTrip);
							if (triplet[0] < tempTrip[0])
								triplet = tempTrip;
							

						}
					}		
				}
			}
		}	
	}
}	

File: go/goldbach.go
package main
//Program to compute Goldbach's Weak Conjecture for input value
//For any odd int > 7, finds triplet of primes w/ smallest vector norm that sum to that value
//By Adrian Faircloth
//09-16-21
import "fmt"
import "math"
import "sort"
 
func isPrime(val int) (int) {
	var isPrime int;
	var i int;
	isPrime = 1;
	if (val == 2) {
		isPrime = 0;
	}
	for i = 2; i*i <= val; i++ {
		if (val % i == 0) {
			isPrime = 0;
		}
	}
	return isPrime;
}

func findTriplet(sum int) ([3]int) {
	//Finds triplet of primes < input whose vector norm is smallest
	//Compares the norm of each triplet found to current smallest norm
	//If current norm < smallest norm, fills triplet array with those values
	//Finally, triplet w/ smallest possible vector norm is stored in array and returned
	//In case of a tie, triplet w/ larger smallest value is returned
	var triplet [3] int;
	var tempTrip [3] int;
	var i int;
	var j int;
	var k int;
	var currNorm float64;
	var smallestNorm float64;
	for i = sum; i > 2; i-- {
		for j = 3; j < sum; j++ {
			if (isPrime(i) == 1 && isPrime(j) == 1) {
				k = sum - (i + j);
				if isPrime(k) == 1 && i != j && i != k && k!= j {
					currNorm = math.Sqrt(float64( (i*i) + (j*j) + (k*k) ));
					if smallestNorm == 0 || currNorm < smallestNorm {
						triplet[0] = i;
						triplet[1] = j;
						triplet[2] = k;
						smallestNorm = currNorm;
						sort.Ints(triplet[:]);
						
			   		} else if currNorm == smallestNorm {
						tempTrip[0] = i;
						tempTrip[1] = j;
						tempTrip[2] = k;
						sort.Ints(tempTrip[:]);
						if triplet[0] < tempTrip[0] {
							triplet = tempTrip;
						}
					}
				}
			} 			
		}  
	}
	return triplet;
}

func main() {
	//Takes user input inpVal and checks if inpVal meets Goldbach requirements
	//If yes, finds triplet w/ smallest norm that sums to inpVal,
	//	sorts triplet and prints
	//If no, prints error message and exits
	var inpVal int;
	fmt.Print("Enter an odd integer greater than seven: ");
	fmt.Scanln(&inpVal);
	var valTriplet [3] int;
	if (inpVal % 2 == 1 && inpVal > 7) {
		valTriplet = findTriplet(inpVal);
		fmt.Println(inpVal, ": (", valTriplet[0], ",", valTriplet[1], ",", valTriplet[2], ")");
	} else {
		fmt.Println("Invalid input (not odd integer greater than seven)");
	}
}
